### PR TITLE
Refinment related update

### PR DESCRIPTION
**Pull request by etola**
### Story:
http://enter-story-URL

### Work done
Add a brief synopsis of the work done, and why they are necessary/wanted. Describe your reasoning behind any opaque decisions here as well. Bonus points for adding comments to the code with links to the PR with the thought process explained.

### Open questions / blockers
- [ ] Any questions that still need to be answered to finish this work (if the PR is a WIP)
- [ ] Titles/descriptions of any work that needs to be done prior to merging this (with links to relevant PRs/stories)

#### Before
Example(s) of output before these changes
#### After
Example(s) of output after these changes

### CODE REVIEW

**Review by kheinzelmann — APPROVED** [review]
LGTM

File: include/kortex/math.h [hunk 1]
```diff
@@ -191,6 +205,13 @@ namespace kortex {
         arr.erase(last, arr.end());
     }
 
+    template<typename T> inline
+    bool is_unique( const vector<T>& arr ) {
+        vector<T> tarr = arr;
+        make_unique(tarr);
```
Comment (gfloros) [inline include/kortex/math.h:211]: Caution! `std::make_unique` makes a pointer, `std::unique_ptr`. To make an `std::vector` unique, you have to `std::sort` it and then use `std::unique`. 
This is a blocking comment!

File: src/image_processing.cc [hunk 2]
```diff
@@ -1693,87 +1666,87 @@ namespace kortex {
...
-			const uchar* brow = B.get_row_u(y);
-			uchar      * crow = C.get_row_u(y);
-			for( int x=0; x<C.w(); x++ ) {
-				if( arow[x] > 0 || brow[x] > 0 ) {
-					crow[x] = 255;
-				}
-			}
-		}
-	}
+    void dilate_image( Image& img, int hsz ) {
```
Comment (gfloros) [inline src/image_processing.cc:1669]: OpenCV may do these faster. And you will also have less code to maintain. 😄 
Not blocking comment, of course.

**Review by gfloros — CHANGES_REQUESTED** [review]
One blocking comment and few non-blocking ones.

File: include/kortex/math.h (hunk 1 above)
Comment (kheinzelmann) [reply include/kortex/math.h:211]: This is calling the `make_unique` function in the kortex namespace. Probably should be a different name but koenig should resolve the correct function.

Comment (etola) [reply include/kortex/math.h:211]: It is actually not calling the `std::make_unique`  but a wrapped function inside the library implementing what you are suggesting in math.h: https://github.com/dronedeploy/kortex/blob/b285321f7eae62f37a837188c85e715e59aa3325/include/kortex/math.h#L188

File: src/image_processing.cc (hunk 2 above)
Comment (etola) [reply src/image_processing.cc:1669]: I didn't want to link kortex with opencv at the time when I started this library to have no external dependency to maintain ;)

File: include/kortex/kvector.h [hunk 3]
```diff
@@ -87,6 +89,25 @@ namespace kortex {
...
+            kortex::clamp_min(m_v, N, vmin);
+        }
+        void max_clamp(const T& vmax) {
+            kortex::clamp_max(m_v, N, vmax);
+        }
+        void clamp(const T& vmin, const T& vmax) {
+            clamp_array(m_v, N, vmin, vmax);
+        }
+        float average() const {
+            return float( std::accumulate(m_v, m_v+N, 0) ) / N;
```
[the remark this answers is not in the document]
Comment (etola) [reply include/kortex/kvector.h:108]: good idea - updating

File: include/kortex/kvector.h [hunk 4]
```diff
@@ -87,6 +89,25 @@ namespace kortex {
             printf( "%f %f %f\n", m_v[0], m_v[1], m_v[2] );
         }
 
+        T max() const {
+            return *std::max_element(m_v, m_v+N);
```
[the remark this answers is not in the document]
Comment (etola) [reply include/kortex/kvector.h:93]: actually with N=0, `std::max_element` doesn't crash as long as m_v is not null - I checked ;) It returns a null pointer if m_v is null and dereferencing it crashes with segfault. Here, `kvector` is a templated class and `N` is its template parameter - > maybe I can enforce N > 0 during template definition but I think that goes into metaprogramming too much - I think I can static_assert N>0 in class definition above though.

**Review by gfloros — COMMENTED** [review]
Thanks for these changes @etola. Could you please also fix the issue with `std::make_unique`?

File: include/kortex/math.h (hunk 1 above)
Comment (etola) [reply include/kortex/math.h:211]: I'll make the namespace explicit : `kortex::make_unique` to avoid confusion

Comment (gfloros) [reply include/kortex/math.h]: Oops, sorry! I didn't see that.

**Review by gfloros — APPROVED** [review]
Thanks for the changes @etola!

**Pull request merged**